numpy-stubs/__init__.pyi

```diff
@@ -87,7 +87,7 @@ _DtypeLike = Union[
 
 _NdArraySubClass = TypeVar("_NdArraySubClass", bound=ndarray)
 
-_ArrayLike = TypeVar("_ArrayLike")
+ArrayLike = Union[int, float, complex, generic, ndarray, Sequence]
```

Comment:
Might it be an idea to, respectively, replace ``int``, ``float`` and ``complex`` with the ``SupportsInt``, ``SupportsFloat`` and ``SupportsComplex`` protocols?

What about ``str``, ``bytes`` and ``dt.datetime`` (considering they all have their corresponding generic)?

Furthermore, I'd presonally be in favor to replace ``ndarray`` with a custom ``_SupportsArray`` protocol along the lines of:
``` python
class _SupportsArray(Protocol):
    def __array__(self, dtype: _DtypeLike = ...) -> ndarray: ...
```

Reply:
> Might it be an idea to, respectively, replace int, float and complex with the SupportsInt, SupportsFloat and SupportsComplex protocols?

I think the protocols are a little too general; e.g.

```
>>> class A:
...     def __int__(self):
...         return 1
...
>>> int(A())
1
>>> np.array(A())
array(<__main__.A object at 0x10e61a290>, dtype=object)
```

> What about str, bytes, bool, dt.datetime dt.date and dt.timedelta (considering they all have their corresponding generic)?

`str` and `bytes` are both covered by `Sequence`. I added bool (thanks for catching that!). The `dt.*` I think also ends up giving unexpected results:

```
>>> np.array(datetime.timedelta(days=1))
array(datetime.timedelta(days=1), dtype=object)
```

> Furthermore, I'd presonally be in favor of replacing ndarray with a custom _SupportsArray protocol along the lines of:

Yes, that is an excellent idea, switched to that.

Reply:
> I think the protocols are a little too general; e.g.

Good catch, I actually wasn't aware that, _e.g._, a ``SupportsInt`` member wouldn't produce an integer array.

> `str` and `bytes` are both covered by `Sequence`

Ah right, that's true.